### ISSUE TITLE
stability: fresh delta cluster experiencing unavailable ranges

### ISSUE THREAD

**Issue by mberhault** [cockroachdb/cockroach#10733]
For the last few days (since we first wiped delta on Nov 14th), every fresh wipe and restart of delta has had issues as soon as block-writer is started.

eg: the latest run was a bit more conservative than before, letting things run for a while.
* started all nodes (from scratch)
* 15 minutes later: start block writer on the first node
* 10 minutes later: start block writer on the second node
* 10 minutes later: start block writer on the third node
* a few minutes later: start block writer on all remaining nodes

The moment the last step was performed, unavailable ranges went to 5 on one of the nodes: `104.196.23.238`

There's a long-running `replicate` op on that node:
```
2016/11/16 15:24:17.409664	1516.572132	replicate
15:24:17.409711	 .    47	... [n9,replicate,s18,r44/1:/Table/51/1/1{446004…-735854…}] processing replica
15:24:17.409723	 .    12	... [n9,replicate,s18,r44/1:/Table/51/1/1{446004…-735854…}] got range lease
15:24:17.409804	 .    81	... [n9,replicate,s18,r44/1:/Table/51/1/1{446004…-735854…}] considering a rebalance
15:24:17.409875	 .    71	... [n9,replicate,s18,r44/1:/Table/51/1/1{446004…-735854…}] rebalancing to {NodeID:1 StoreID:1 ReplicaID:0}
15:24:17.409890	 .    15	... [n9,replicate,s18,r44/1:/Table/51/1/1{446004…-735854…}] snapshot retry loop pass 0
15:24:17.409901	 .    12	... [n9,replicate,s18,r44/1:/Table/51/1/1{446004…-735854…}] snapshot already running
15:24:17.409911	 .     9	... [n9,replicate,s18,r44/1:/Table/51/1/1{446004…-735854…}] snapshot retry loop pass 1
15:24:17.409919	 .     8	... [n9,replicate,s18,r44/1:/Table/51/1/1{446004…-735854…}] snapshot already running
15:24:17.409929	 .    10	... [n9,replicate,s18,r44/1:/Table/51/1/1{446004…-735854…}] snapshot retry loop pass 2
15:24:17.409942	 .    13	... [n9,replicate,s18,r44/1:/Table/51/1/1{446004…-735854…}] snapshot already running
15:24:17.409949	 .     7	... [n9,replicate,s18,r44/1:/Table/51/1/1{446004…-735854…}] snapshot retry loop pass 3
15:24:17.409958	 .     8	... [n9,replicate,s18,r44/1:/Table/51/1/1{446004…-735854…}] snapshot already running
15:24:17.409965	 .     8	... [n9,replicate,s18,r44/1:/Table/51/1/1{446004…-735854…}] snapshot retry loop pass 4
15:24:17.409992	 .    27	... [n9,replicate,s18,r44/1:/Table/51/1/1{446004…-735854…}] snapshot already running
...
15:49:34.008687	 .     9	... [n9,replicate,s18,r44/1:/Table/51/1/1{446004…-735854…}] snapshot already running
15:49:34.008726	 .     6	... [n9,replicate,s18,r44/1:/Table/51/1/1{446004…-735854…}] snapshot already running
15:49:34.008742	 .     8	... [n9,replicate,s18,r44/1:/Table/51/1/1{446004…-735854…}] snapshot already running
15:49:34.008772	 .     8	... [n9,replicate,s18,r44/1:/Table/51/1/1{446004…-735854…}] snapshot retry loop pass 87437375
```

This matches the time when the last set of block_writers were started.
This is nodeID 9.

Looking at its `/debug/raft` page, that ranges shows:
```
    "44": {
      "rangeId": "44",
      "errors": [
        {
          "message": "node 10 range descriptor does not match node 9"
        }
      ],
      "nodes": [
        {
          "nodeId": 9,
          "range": {
            "span": {
              "startKey": "/Table/51/1/1446004355585826718/\"ea8c1c27-5240-4764-9428-22561a6b10bf\"/100751",
              "endKey": "/Table/51/1/1735854988129937498/\"12b68893-1568-425f-8d2f-3e3e441ec6ce\"/30880"
            },
            "raftState": {
              "replicaId": "1",
              "hardState": {
                "term": "7",
                "vote": "1",
                "commit": "234337"
              },
              "lead": "1",
              "state": "StateLeader",
              "applied": "234271",
              "progress": {
                "1": {
                  "match": "234337",
                  "next": "234338",
                  "state": "ProgressStateProbe",
                  "paused": false,
                  "pendingSnapshot": "0"
                },
                "16": {
                  "match": "234335",
                  "next": "234338",
                  "state": "ProgressStateReplicate",
                  "paused": false,
                  "pendingSnapshot": "0"
                },
                "17": {
                  "match": "234337",
                  "next": "234338",
                  "state": "ProgressStateReplicate",
                  "paused": false,
                  "pendingSnapshot": "0"
                }
              }
            },
            "state": {
              "state": {
                "raftAppliedIndex": "234275",
                "leaseAppliedIndex": "234143",
                "desc": {
                  "rangeId": "44",
                  "startKey": "u4n9FBE9TNZcX54SZWE4YzFjMjctNTI0MC00NzY0LTk0MjgtMjI1NjFhNmIxMGJmAAH4AYmP",
                  "endKey": "u4n9GBb+6i0TKFoSMTJiNjg4OTMtMTU2OC00MjVmLThkMmYtM2UzZTQ0MWVjNmNlAAH3eKA=",
                  "replicas": [
                    {
                      "nodeId": 9,
                      "storeId": 18,
                      "replicaId": 1
                    },
                    {
                      "nodeId": 3,
                      "storeId": 6,
                      "replicaId": 16
                    },
                    {
                      "nodeId": 4,
                      "storeId": 8,
                      "replicaId": 17
                    }
                  ],
                  "nextReplicaId": 18
                },
                "lease": {
                  "start": {
                    "wallTime": "1479309084380384195",
                    "logical": 0
                  },
                  "startStasis": {
                    "wallTime": "1479309873076966273",
                    "logical": 0
                  },
                  "expiration": {
                    "wallTime": "1479309873326966273",
                    "logical": 0
                  },
                  "replica": {
                    "nodeId": 9,
                    "storeId": 18,
                    "replicaId": 1
                  },
                  "proposedTs": {
                    "wallTime": "1479309864076972237",
                    "logical": 0
                  }
                },
                "truncatedState": {
                  "index": "231264",
                  "term": "7"
                },
                "gcThreshold": {
                  "wallTime": "0",
                  "logical": 0
                },
                "stats": {
                  "containsEstimates": false,
                  "lastUpdateNanos": "1479309864673732033",
                  "intentAge": "782",
                  "gcBytesAge": "0",
                  "liveBytes": "51600565",
                  "liveCount": "363407",
                  "keyBytes": "24489799",
                  "keyCount": "363407",
                  "valBytes": "27110766",
                  "valCount": "363407",
                  "intentBytes": "17556",
                  "intentCount": "201",
                  "sysBytes": "30861",
                  "sysCount": "50"
                },
                "txnSpanGcThreshold": {
                  "wallTime": "0",
                  "logical": 0
                },
                "frozen": 2
              },
              "lastIndex": "234337",
              "numPending": "62",
              "numDropped": "0",
              "raftLogSize": "4157078"
            }
          }
        },
        {
          "nodeId": 10,
          "range": {
            "span": {
              "startKey": "/Table/51/1/1446004355585826718/\"ea8c1c27-5240-4764-9428-22561a6b10bf\"/100751",
              "endKey": "/Table/51/1/1591216144711479990/\"56b80c99-42d2-434b-8a8f-b612273517db\"/123483"
            },
            "raftState": {
              "replicaId": "18",
              "hardState": {
                "term": "13",
                "vote": "20",
                "commit": "470943"
              },
              "lead": "20",
              "state": "StateFollower",
              "applied": "470943"
            },
            "state": {
              "state": {
                "raftAppliedIndex": "470943",
                "leaseAppliedIndex": "470591",
                "desc": {
                  "rangeId": "44",
                  "startKey": "u4n9FBE9TNZcX54SZWE4YzFjMjctNTI0MC00NzY0LTk0MjgtMjI1NjFhNmIxMGJmAAH4AYmP",
                  "endKey": "u4n9FhUipypk8rYSNTZiODBjOTktNDJkMi00MzRiLThhOGYtYjYxMjI3MzUxN2RiAAH4AeJb",
                  "replicas": [
                    {
                      "nodeId": 10,
                      "storeId": 20,
                      "replicaId": 18
                    },
                    {
                      "nodeId": 7,
                      "storeId": 13,
                      "replicaId": 20
                    },
                    {
                      "nodeId": 4,
                      "storeId": 8,
                      "replicaId": 17
                    }
                  ],
                  "nextReplicaId": 22
                },
                "lease": {
                  "start": {
                    "wallTime": "1479311378880480247",
                    "logical": 1
                  },
                  "startStasis": {
                    "wallTime": "1479311389924558799",
                    "logical": 0
                  },
                  "expiration": {
                    "wallTime": "1479311390174558799",
                    "logical": 0
                  },
                  "replica": {
                    "nodeId": 7,
                    "storeId": 13,
                    "replicaId": 20
                  },
                  "proposedTs": {
                    "wallTime": "1479311380924570592",
                    "logical": 0
                  }
                },
                "truncatedState": {
                  "index": "470938",
                  "term": "13"
                },
                "gcThreshold": {
                  "wallTime": "0",
                  "logical": 0
                },
                "stats": {
                  "containsEstimates": false,
                  "lastUpdateNanos": "1479311381761426084",
                  "intentAge": "343837",
                  "gcBytesAge": "0",
                  "liveBytes": "36150276",
                  "liveCount": "254388",
                  "keyBytes": "17170701",
                  "keyCount": "254388",
                  "valBytes": "18979575",
                  "valCount": "254388",
                  "intentBytes": "22419",
                  "intentCount": "262",
                  "sysBytes": "39413",
                  "sysCount": "55"
                },
                "txnSpanGcThreshold": {
                  "wallTime": "0",
                  "logical": 0
                },
                "frozen": 2
              },
              "lastIndex": "470943",
              "numPending": "0",
              "numDropped": "0",
              "raftLogSize": "70021836"
            }
          }
        },
        {
          "nodeId": 7,
          "range": {
            "span": {
              "startKey": "/Table/51/1/1446004355585826718/\"ea8c1c27-5240-4764-9428-22561a6b10bf\"/100751",
              "endKey": "/Table/51/1/1591216144711479990/\"56b80c99-42d2-434b-8a8f-b612273517db\"/123483"
            },
            "raftState": {
              "replicaId": "20",
              "hardState": {
                "term": "13",
                "vote": "20",
                "commit": "470943"
              },
              "lead": "20",
              "state": "StateLeader",
              "applied": "470943",
              "progress": {
                "17": {
                  "match": "470943",
                  "next": "470944",
                  "state": "ProgressStateReplicate",
                  "paused": false,
                  "pendingSnapshot": "0"
                },
                "18": {
                  "match": "470943",
                  "next": "470944",
                  "state": "ProgressStateReplicate",
                  "paused": false,
                  "pendingSnapshot": "0"
                },
                "20": {
                  "match": "470943",
                  "next": "470944",
                  "state": "ProgressStateProbe",
                  "paused": false,
                  "pendingSnapshot": "0"
                }
              }
            },
            "state": {
              "state": {
                "raftAppliedIndex": "470943",
                "leaseAppliedIndex": "470591",
                "desc": {
                  "rangeId": "44",
                  "startKey": "u4n9FBE9TNZcX54SZWE4YzFjMjctNTI0MC00NzY0LTk0MjgtMjI1NjFhNmIxMGJmAAH4AYmP",
                  "endKey": "u4n9FhUipypk8rYSNTZiODBjOTktNDJkMi00MzRiLThhOGYtYjYxMjI3MzUxN2RiAAH4AeJb",
                  "replicas": [
                    {
                      "nodeId": 10,
                      "storeId": 20,
                      "replicaId": 18
                    },
                    {
                      "nodeId": 7,
                      "storeId": 13,
                      "replicaId": 20
                    },
                    {
                      "nodeId": 4,
                      "storeId": 8,
                      "replicaId": 17
                    }
                  ],
                  "nextReplicaId": 22
                },
                "lease": {
                  "start": {
                    "wallTime": "1479311378880480247",
                    "logical": 1
                  },
                  "startStasis": {
                    "wallTime": "1479311389924558799",
                    "logical": 0
                  },
                  "expiration": {
                    "wallTime": "1479311390174558799",
                    "logical": 0
                  },
                  "replica": {
                    "nodeId": 7,
                    "storeId": 13,
                    "replicaId": 20
                  },
                  "proposedTs": {
                    "wallTime": "1479311380924570592",
                    "logical": 0
                  }
                },
                "truncatedState": {
                  "index": "470938",
                  "term": "13"
                },
                "gcThreshold": {
                  "wallTime": "0",
                  "logical": 0
                },
                "stats": {
                  "containsEstimates": false,
                  "lastUpdateNanos": "1479311381761426084",
                  "intentAge": "343837",
                  "gcBytesAge": "0",
                  "liveBytes": "36150276",
                  "liveCount": "254388",
                  "keyBytes": "17170701",
                  "keyCount": "254388",
                  "valBytes": "18979575",
                  "valCount": "254388",
                  "intentBytes": "22419",
                  "intentCount": "262",
                  "sysBytes": "39413",
                  "sysCount": "55"
                },
                "txnSpanGcThreshold": {
                  "wallTime": "0",
                  "logical": 0
                },
                "frozen": 2
              },
              "lastIndex": "470943",
              "numPending": "0",
              "numDropped": "0",
              "raftLogSize": "11677207"
            }
          }
        },
        {
          "nodeId": 4,
          "range": {
            "span": {
              "startKey": "/Table/51/1/1446004355585826718/\"ea8c1c27-5240-4764-9428-22561a6b10bf\"/100751",
              "endKey": "/Table/51/1/1591216144711479990/\"56b80c99-42d2-434b-8a8f-b612273517db\"/123483"
            },
            "raftState": {
              "replicaId": "17",
              "hardState": {
                "term": "13",
                "vote": "20",
                "commit": "470943"
              },
              "lead": "20",
              "state": "StateFollower",
              "applied": "470943"
            },
            "state": {
              "state": {
                "raftAppliedIndex": "470943",
                "leaseAppliedIndex": "470591",
                "desc": {
                  "rangeId": "44",
                  "startKey": "u4n9FBE9TNZcX54SZWE4YzFjMjctNTI0MC00NzY0LTk0MjgtMjI1NjFhNmIxMGJmAAH4AYmP",
                  "endKey": "u4n9FhUipypk8rYSNTZiODBjOTktNDJkMi00MzRiLThhOGYtYjYxMjI3MzUxN2RiAAH4AeJb",
                  "replicas": [
                    {
                      "nodeId": 10,
                      "storeId": 20,
                      "replicaId": 18
                    },
                    {
                      "nodeId": 7,
                      "storeId": 13,
                      "replicaId": 20
                    },
                    {
                      "nodeId": 4,
                      "storeId": 8,
                      "replicaId": 17
                    }
                  ],
                  "nextReplicaId": 22
                },
                "lease": {
                  "start": {
                    "wallTime": "1479311378880480247",
                    "logical": 1
                  },
                  "startStasis": {
                    "wallTime": "1479311389924558799",
                    "logical": 0
                  },
                  "expiration": {
                    "wallTime": "1479311390174558799",
                    "logical": 0
                  },
                  "replica": {
                    "nodeId": 7,
                    "storeId": 13,
                    "replicaId": 20
                  },
                  "proposedTs": {
                    "wallTime": "1479311380924570592",
                    "logical": 0
                  }
                },
                "truncatedState": {
                  "index": "470938",
                  "term": "13"
                },
                "gcThreshold": {
                  "wallTime": "0",
                  "logical": 0
                },
                "stats": {
                  "containsEstimates": false,
                  "lastUpdateNanos": "1479311381761426084",
                  "intentAge": "343837",
                  "gcBytesAge": "0",
                  "liveBytes": "36150276",
                  "liveCount": "254388",
                  "keyBytes": "17170701",
                  "keyCount": "254388",
                  "valBytes": "18979575",
                  "valCount": "254388",
                  "intentBytes": "22419",
                  "intentCount": "262",
                  "sysBytes": "39413",
                  "sysCount": "55"
                },
                "txnSpanGcThreshold": {
                  "wallTime": "0",
                  "logical": 0
                },
                "frozen": 2
              },
              "lastIndex": "470943",
              "numPending": "0",
              "numDropped": "0",
              "raftLogSize": "59378075"
            }
          }
        }
      ]
    },
```

Some odd details include:
```
                      "nodeId": 3,
"nodeId": 9,
    "startKey": "/Table/51/1/1446004355585826718/\"ea8c1c27-5240-4764-9428-22561a6b10bf\"/100751",
    "endKey": "/Table/51/1/1735854988129937498/\"12b68893-1568-425f-8d2f-3e3e441ec6ce\"/30880"
    "state": "StateLeader",
"nodeId": 10,
    "startKey": "/Table/51/1/1446004355585826718/\"ea8c1c27-5240-4764-9428-22561a6b10bf\"/100751",
    "endKey": "/Table/51/1/1591216144711479990/\"56b80c99-42d2-434b-8a8f-b612273517db\"/123483"
    "state": "StateFollower",
"nodeId": 7,
    "startKey": "/Table/51/1/1446004355585826718/\"ea8c1c27-5240-4764-9428-22561a6b10bf\"/100751",
    "endKey": "/Table/51/1/1591216144711479990/\"56b80c99-42d2-434b-8a8f-b612273517db\"/123483"
    "state": "StateLeader",
"nodeId": 4,
    "startKey": "/Table/51/1/1446004355585826718/\"ea8c1c27-5240-4764-9428-22561a6b10bf\"/100751",
    "endKey": "/Table/51/1/1591216144711479990/\"56b80c99-42d2-434b-8a8f-b612273517db\"/123483"
    "state": "StateFollower",
```

This could very well just be me being confused about what this should really be showing, but it certainly looks like two nodes think they're leaders of this range.

**Comment by bdarnell** [comment]
Two leaders for a short time is not necessarily a problem (as long as they're at different terms, which we can't tell from here); it just takes time for the previous leader to find out it's been deposed. (see #6409 for a past instance of this, which we believe has been fixed). This could be caused by something blocking the raft goroutines for too long (which is likely given how much we've been seeing the mutex warnings in the logs, although I haven't looked at these logs on delta yet)

**Comment by mberhault** [comment]
I tried delta again with a single store per node since we were confused about this not happening on other clusters.
Not only did this happen again, but as I wiped and restarted blue, it ran into the same symptoms.

**Comment by bdarnell** [comment]
There are several interesting things about that retry loop. First of all, it's run 80 million times; I've put a stop to that in #10753. Second, the last several iterations have "snapshot already running" without matching "snapshot retry loop pass" entries. I don't see how that's possible with the way GetSnapshot (which prints the retry loop message) calls SnapshotWithContext (which prints the "already running" message).

Ignoring that for a moment, we know that this is a preemptive snapshot, since that is the only non-test place where we call GetSnapshot. Only one thread (the replication queue) ever generates preemptive snapshots, so the snapshot that is "already running" is either A) leaked by an earlier iteration of the replication queue or B) a raft-generated snapshot. The latter is much more likely to leak since there we just defer the call to CloseOutSnap in the replication queue while the relationship is much more subtle for raft snapshots.

**Comment by bdarnell** [comment]
Looking for snapshot-related stack traces, on one delta node I find these:

```
173 @ 0x631faa 0x63209e 0x64221d 0x641eb0 0x6de5a0 0xe08d60 0xb36452 0xb36271 0xb31b05 0xb30576 0xb4b11f 0xb4b39d 0x7241af 0x663351
#	0x641eaf	sync.runtime_Semacquire+0x2f									/usr/local/go/src/runtime/sema.go:47
#	0x6de59f	sync.(*Mutex).Lock+0xcf										/usr/local/go/src/sync/mutex.go:85
#	0xe08d5f	github.com/cockroachdb/cockroach/pkg/util/syncutil.(*TimedMutex).Lock+0x2f			/go/src/github.com/cockroachdb/cockroach/pkg/util/syncutil/timedmutex.go:84
#	0xb36451	github.com/cockroachdb/cockroach/pkg/storage.(*Store).tryGetOrCreateReplica+0xf1		/go/src/github.com/cockroachdb/cockroach/pkg/storage/store.go:3443
#	0xb36270	github.com/cockroachdb/cockroach/pkg/storage.(*Store).getOrCreateReplica+0x50			/go/src/github.com/cockroachdb/cockroach/pkg/storage/store.go:3407
#	0xb31b04	github.com/cockroachdb/cockroach/pkg/storage.(*Store).processRaftRequest+0x94			/go/src/github.com/cockroachdb/cockroach/pkg/storage/store.go:2721
#	0xb30575	github.com/cockroachdb/cockroach/pkg/storage.(*Store).HandleSnapshot+0x745			/go/src/github.com/cockroachdb/cockroach/pkg/storage/store.go:2608
#	0xb4b11e	github.com/cockroachdb/cockroach/pkg/storage.(*RaftTransport).RaftSnapshot.func1.1+0x17e	/go/src/github.com/cockroachdb/cockroach/pkg/storage/raft_transport.go:362
#	0xb4b39c	github.com/cockroachdb/cockroach/pkg/storage.(*RaftTransport).RaftSnapshot.func1+0x5c		/go/src/github.com/cockroachdb/cockroach/pkg/storage/raft_transport.go:363
#	0x7241ae	github.com/cockroachdb/cockroach/pkg/util/stop.(*Stopper).RunAsyncTask.func1+0xde		/go/src/github.com/cockroachdb/cockroach/pkg/util/stop/stopper.go:264

173 @ 0x631faa 0x6416a9 0x64049c 0xad0d3b 0xac351b 0xd27172 0xd28086 0xd2c88b 0x663351
#	0xad0d3a	github.com/cockroachdb/cockroach/pkg/storage.(*RaftTransport).RaftSnapshot+0x23a	/go/src/github.com/cockroachdb/cockroach/pkg/storage/raft_transport.go:367
#	0xac351a	github.com/cockroachdb/cockroach/pkg/storage._MultiRaft_RaftSnapshot_Handler+0xba	/go/src/github.com/cockroachdb/cockroach/pkg/storage/raft.pb.go:353
#	0xd27171	google.golang.org/grpc.(*Server).processStreamingRPC+0x701				/go/src/google.golang.org/grpc/server.go:689
#	0xd28085	google.golang.org/grpc.(*Server).handleStream+0xc35					/go/src/google.golang.org/grpc/server.go:773
#	0xd2c88a	google.golang.org/grpc.(*Server).serveStreams.func1.1+0xaa				/go/src/google.golang.org/grpc/server.go:422
```

All server-side calls blocked acquiring a mutex. This is one way that we could see "leaked" snapshots - it's not completely leaked, but it's waiting on something with no timeouts. The mutex they're waiting for is the replica's `raftMu`.

We have a couple of functions named `RaftMuLocked`, and we see a couple of threads running them:

```
2 @ 0x631faa 0x6416a9 0x64049c 0x722cc6 0xaa5e7a 0xb0e520 0xb0ec32 0xae4f02 0xadfd29 0xadf42d 0xb3509e 0xb2027c 0xb559a3 0x72408d 0x663351
#	0x722cc5	github.com/cockroachdb/cockroach/pkg/util/stop.(*Stopper).RunLimitedAsyncTask+0x855		/go/src/github.com/cockroachdb/cockroach/pkg/util/stop/stopper.go:296
#	0xaa5e79	github.com/cockroachdb/cockroach/pkg/storage.(*intentResolver).processIntentsAsync+0x479	/go/src/github.com/cockroachdb/cockroach/pkg/storage/intent_resolver.go:383
#	0xb0e51f	github.com/cockroachdb/cockroach/pkg/storage.(*Replica).handleLocalProposalData+0x10f		/go/src/github.com/cockroachdb/cockroach/pkg/storage/replica_proposal.go:603
#	0xb0ec31	github.com/cockroachdb/cockroach/pkg/storage.(*Replica).handleProposalData+0x111		/go/src/github.com/cockroachdb/cockroach/pkg/storage/replica_proposal.go:679
#	0xae4f01	github.com/cockroachdb/cockroach/pkg/storage.(*Replica).processRaftCommand+0x701		/go/src/github.com/cockroachdb/cockroach/pkg/storage/replica.go:3190
#	0xadfd28	github.com/cockroachdb/cockroach/pkg/storage.(*Replica).handleRaftReadyRaftMuLocked+0x898	/go/src/github.com/cockroachdb/cockroach/pkg/storage/replica.go:2351
#	0xadf42c	github.com/cockroachdb/cockroach/pkg/storage.(*Replica).handleRaftReady+0xcc			/go/src/github.com/cockroachdb/cockroach/pkg/storage/replica.go:2179
#	0xb3509d	github.com/cockroachdb/cockroach/pkg/storage.(*Store).processReady+0x13d			/go/src/github.com/cockroachdb/cockroach/pkg/storage/store.go:3201
#	0xb2027b	github.com/cockroachdb/cockroach/pkg/storage.(*raftScheduler).worker+0x28b			/go/src/github.com/cockroachdb/cockroach/pkg/storage/scheduler.go:230
#	0xb559a2	github.com/cockroachdb/cockroach/pkg/storage.(*raftScheduler).Start.func2+0x32			/go/src/github.com/cockroachdb/cockroach/pkg/storage/scheduler.go:181
#	0x72408c	github.com/cockroachdb/cockroach/pkg/util/stop.(*Stopper).RunWorker.func1+0x7c			/go/src/github.com/cockroachdb/cockroach/pkg/util/stop/stopper.go:196
```

They're both blocked in RunLimitedAsyncTask. This is the backpressure that is supposed to slow down clients if too much intent-resolution work is being queued up, but it's possible that it's getting into some sort of deadlock.

We have 100 threads (with two different stack traces) that descend from intent_resolver.go:378, consuming slots in intentResolverTaskLimit. 17 of them are waiting on raftMu, which partially explains the deadlock. It's a lock-order violation to be sure, but it's unclear what the other 83 are waiting on. They're sitting in the select loop that was just changed in #10279; my guess is that bisection would point to that commit. 

```
17 @ 0x631faa 0x63209e 0x64221d 0x641eb0 0x6de5a0 0xe08d60 0xadde24 0xadc622 0xadc44a 0xad7c82 0xb2eb3e 0xb3d362 0xa8a651 0x722150 0xa76ffd 0xa772fb 0x1028875 0x1020834 0x101b604 0x101bc18 0x101e88f 0x101d52c 0x101c893 0x102afe9 0xc510bf 0xc5750e 0xc505e8 0xc508cc 0xb47a3b 0x72436c 0x663351
#	0x641eaf	sync.runtime_Semacquire+0x2f									/usr/local/go/src/runtime/sema.go:47
#	0x6de59f	sync.(*Mutex).Lock+0xcf										/usr/local/go/src/sync/mutex.go:85
#	0xe08d5f	github.com/cockroachdb/cockroach/pkg/util/syncutil.(*TimedMutex).Lock+0x2f			/go/src/github.com/cockroachdb/cockroach/pkg/util/syncutil/timedmutex.go:84
#	0xadde23	github.com/cockroachdb/cockroach/pkg/storage.(*Replica).propose+0x1e3				/go/src/github.com/cockroachdb/cockroach/pkg/storage/replica.go:1975
#	0xadc621	github.com/cockroachdb/cockroach/pkg/storage.(*Replica).tryAddWriteCmd+0x181			/go/src/github.com/cockroachdb/cockroach/pkg/storage/replica.go:1782
#	0xadc449	github.com/cockroachdb/cockroach/pkg/storage.(*Replica).addWriteCmd+0x79			/go/src/github.com/cockroachdb/cockroach/pkg/storage/replica.go:1687
#	0xad7c81	github.com/cockroachdb/cockroach/pkg/storage.(*Replica).Send+0x221				/go/src/github.com/cockroachdb/cockroach/pkg/storage/replica.go:1142
#	0xb2eb3d	github.com/cockroachdb/cockroach/pkg/storage.(*Store).Send+0x75d				/go/src/github.com/cockroachdb/cockroach/pkg/storage/store.go:2372
#	0xb3d361	github.com/cockroachdb/cockroach/pkg/storage.(*Stores).Send+0x201				/go/src/github.com/cockroachdb/cockroach/pkg/storage/stores.go:187
#	0xa8a650	github.com/cockroachdb/cockroach/pkg/server.(*Node).batchInternal.func1+0x2a0			/go/src/github.com/cockroachdb/cockroach/pkg/server/node.go:843
#	0x72214f	github.com/cockroachdb/cockroach/pkg/util/stop.(*Stopper).RunTaskWithErr+0xff			/go/src/github.com/cockroachdb/cockroach/pkg/util/stop/stopper.go:245
#	0xa76ffc	github.com/cockroachdb/cockroach/pkg/server.(*Node).batchInternal+0x1ec				/go/src/github.com/cockroachdb/cockroach/pkg/server/node.go:854
#	0xa772fa	github.com/cockroachdb/cockroach/pkg/server.(*Node).Batch+0x9a					/go/src/github.com/cockroachdb/cockroach/pkg/server/node.go:874
#	0x1028874	github.com/cockroachdb/cockroach/pkg/kv.(*grpcTransport).SendNext+0x1c4				/go/src/github.com/cockroachdb/cockroach/pkg/kv/transport.go:184
#	0x1020833	github.com/cockroachdb/cockroach/pkg/kv.(*DistSender).sendToReplicas+0x353			/go/src/github.com/cockroachdb/cockroach/pkg/kv/dist_sender.go:1083
#	0x101b603	github.com/cockroachdb/cockroach/pkg/kv.(*DistSender).sendRPC+0x163				/go/src/github.com/cockroachdb/cockroach/pkg/kv/dist_sender.go:364
#	0x101bc17	github.com/cockroachdb/cockroach/pkg/kv.(*DistSender).sendSingleRange+0x147			/go/src/github.com/cockroachdb/cockroach/pkg/kv/dist_sender.go:433
#	0x101e88e	github.com/cockroachdb/cockroach/pkg/kv.(*DistSender).sendPartialBatch+0x29e			/go/src/github.com/cockroachdb/cockroach/pkg/kv/dist_sender.go:874
#	0x101d52b	github.com/cockroachdb/cockroach/pkg/kv.(*DistSender).divideAndSendBatchToRanges+0x41b		/go/src/github.com/cockroachdb/cockroach/pkg/kv/dist_sender.go:756
#	0x101c892	github.com/cockroachdb/cockroach/pkg/kv.(*DistSender).Send+0x2b2				/go/src/github.com/cockroachdb/cockroach/pkg/kv/dist_sender.go:577
#	0x102afe8	github.com/cockroachdb/cockroach/pkg/kv.(*TxnCoordSender).Send+0x7e8				/go/src/github.com/cockroachdb/cockroach/pkg/kv/txn_coord_sender.go:416
#	0xc510be	github.com/cockroachdb/cockroach/pkg/internal/client.(*DB).send+0xbe				/go/src/github.com/cockroachdb/cockroach/pkg/internal/client/db.go:511
#	0xc5750d	github.com/cockroachdb/cockroach/pkg/internal/client.(*DB).Run.func1+0x6d			/go/src/github.com/cockroachdb/cockroach/pkg/internal/client/db.go:445
#	0xc505e7	github.com/cockroachdb/cockroach/pkg/internal/client.sendAndFill+0x167				/go/src/github.com/cockroachdb/cockroach/pkg/internal/client/db.go:416
#	0xc508cb	github.com/cockroachdb/cockroach/pkg/internal/client.(*DB).Run+0xab				/go/src/github.com/cockroachdb/cockroach/pkg/internal/client/db.go:447
#	0xb47a3a	github.com/cockroachdb/cockroach/pkg/storage.(*intentResolver).processIntentsAsync.func2+0x63a	/go/src/github.com/cockroachdb/cockroach/pkg/storage/intent_resolver.go:378
#	0x72436b	github.com/cockroachdb/cockroach/pkg/util/stop.(*Stopper).RunLimitedAsyncTask.func1+0x10b	/go/src/github.com/cockroachdb/cockroach/pkg/util/stop/stopper.go:326

83 @ 0x631faa 0x6416a9 0x64049c 0xadca00 0xadc44a 0xad7c82 0xb2eb3e 0xb3d362 0xa8a651 0x722150 0xa76ffd 0xa772fb 0x1028875 0x1020834 0x101b604 0x101bc18 0x101e88f 0x101d52c 0x101c893 0x102afe9 0xc510bf 0xc5750e 0xc505e8 0xc508cc 0xb47a3b 0x72436c 0x663351
#	0xadc9ff	github.com/cockroachdb/cockroach/pkg/storage.(*Replica).tryAddWriteCmd+0x55f			/go/src/github.com/cockroachdb/cockroach/pkg/storage/replica.go:1791
#	0xadc449	github.com/cockroachdb/cockroach/pkg/storage.(*Replica).addWriteCmd+0x79			/go/src/github.com/cockroachdb/cockroach/pkg/storage/replica.go:1687
#	0xad7c81	github.com/cockroachdb/cockroach/pkg/storage.(*Replica).Send+0x221				/go/src/github.com/cockroachdb/cockroach/pkg/storage/replica.go:1142
#	0xb2eb3d	github.com/cockroachdb/cockroach/pkg/storage.(*Store).Send+0x75d				/go/src/github.com/cockroachdb/cockroach/pkg/storage/store.go:2372
#	0xb3d361	github.com/cockroachdb/cockroach/pkg/storage.(*Stores).Send+0x201				/go/src/github.com/cockroachdb/cockroach/pkg/storage/stores.go:187
#	0xa8a650	github.com/cockroachdb/cockroach/pkg/server.(*Node).batchInternal.func1+0x2a0			/go/src/github.com/cockroachdb/cockroach/pkg/server/node.go:843
#	0x72214f	github.com/cockroachdb/cockroach/pkg/util/stop.(*Stopper).RunTaskWithErr+0xff			/go/src/github.com/cockroachdb/cockroach/pkg/util/stop/stopper.go:245
#	0xa76ffc	github.com/cockroachdb/cockroach/pkg/server.(*Node).batchInternal+0x1ec				/go/src/github.com/cockroachdb/cockroach/pkg/server/node.go:854
#	0xa772fa	github.com/cockroachdb/cockroach/pkg/server.(*Node).Batch+0x9a					/go/src/github.com/cockroachdb/cockroach/pkg/server/node.go:874
#	0x1028874	github.com/cockroachdb/cockroach/pkg/kv.(*grpcTransport).SendNext+0x1c4				/go/src/github.com/cockroachdb/cockroach/pkg/kv/transport.go:184
#	0x1020833	github.com/cockroachdb/cockroach/pkg/kv.(*DistSender).sendToReplicas+0x353			/go/src/github.com/cockroachdb/cockroach/pkg/kv/dist_sender.go:1083
#	0x101b603	github.com/cockroachdb/cockroach/pkg/kv.(*DistSender).sendRPC+0x163				/go/src/github.com/cockroachdb/cockroach/pkg/kv/dist_sender.go:364
#	0x101bc17	github.com/cockroachdb/cockroach/pkg/kv.(*DistSender).sendSingleRange+0x147			/go/src/github.com/cockroachdb/cockroach/pkg/kv/dist_sender.go:433
#	0x101e88e	github.com/cockroachdb/cockroach/pkg/kv.(*DistSender).sendPartialBatch+0x29e			/go/src/github.com/cockroachdb/cockroach/pkg/kv/dist_sender.go:874
#	0x101d52b	github.com/cockroachdb/cockroach/pkg/kv.(*DistSender).divideAndSendBatchToRanges+0x41b		/go/src/github.com/cockroachdb/cockroach/pkg/kv/dist_sender.go:756
#	0x101c892	github.com/cockroachdb/cockroach/pkg/kv.(*DistSender).Send+0x2b2				/go/src/github.com/cockroachdb/cockroach/pkg/kv/dist_sender.go:577
#	0x102afe8	github.com/cockroachdb/cockroach/pkg/kv.(*TxnCoordSender).Send+0x7e8				/go/src/github.com/cockroachdb/cockroach/pkg/kv/txn_coord_sender.go:416
#	0xc510be	github.com/cockroachdb/cockroach/pkg/internal/client.(*DB).send+0xbe				/go/src/github.com/cockroachdb/cockroach/pkg/internal/client/db.go:511
#	0xc5750d	github.com/cockroachdb/cockroach/pkg/internal/client.(*DB).Run.func1+0x6d			/go/src/github.com/cockroachdb/cockroach/pkg/internal/client/db.go:445
#	0xc505e7	github.com/cockroachdb/cockroach/pkg/internal/client.sendAndFill+0x167				/go/src/github.com/cockroachdb/cockroach/pkg/internal/client/db.go:416
#	0xc508cb	github.com/cockroachdb/cockroach/pkg/internal/client.(*DB).Run+0xab				/go/src/github.com/cockroachdb/cockroach/pkg/internal/client/db.go:447
#	0xb47a3a	github.com/cockroachdb/cockroach/pkg/storage.(*intentResolver).processIntentsAsync.func2+0x63a	/go/src/github.com/cockroachdb/cockroach/pkg/storage/intent_resolver.go:378
#	0x72436b	github.com/cockroachdb/cockroach/pkg/util/stop.(*Stopper).RunLimitedAsyncTask.func1+0x10b	/go/src/github.com/cockroachdb/cockroach/pkg/util/stop/stopper.go:326
```

**Comment by a-robinson** [comment]
You've probably already noticed, but since it hasn't been mentioned, the replica on node 9 is for a larger range of data. Its end key is "/Table/51/1/17..." vs the end key of ""/Table/51/1/15" that the other replicas have, so it looks like there was a range split around this time.

I'm just taking a complete shot in the dark here, but maybe there's some bad interaction between snapshots and splits?

**Comment by tamird** [comment]
Meta comment: is it time to run a cluster with the `deadlock` tag?

**Comment by petermattis** [comment]
Seems like this is still happening.

**Comment by petermattis** [comment]
From another issue:

I reset `delta` and pushed a build containing Spencer's fix and we're still seeing problems, though the issue appears a bit different. `/debug/requests` shows:

```
2016/11/17 20:38:00.399177	234.495979	node.Batch
20:38:00.399186	 .     9	... txnID:acfaec66-07ab-4682-96f5-e3d94895f6a5
20:38:00.399190	 .     4	... node.Batch
20:38:00.399202	 .    13	... [n1] 1 EndTxn
20:38:00.399215	 .    13	... [n1,s1] executing 1 requests
20:38:00.399238	 .    23	... [n1,s1,r22/1:/Table/51{-/1/46064…}] read-write path
20:38:00.399286	 .    48	... [n1,s1,r22/1:/Table/51{-/1/46064…}] command queue
20:38:00.399570	 .   284	... [n1,s1,r22/1:/Table/51{-/1/46064…}] raft
20:38:39.325228	38.925658	... [n1,s1,r22/1:/Table/51{-/1/46064…}] re-submitting command 6bd387b7c8370ea0 to Raft: reasonNewLeader
20:38:42.686559	3.361331	... [n1,s1,r22/1:/Table/51{-/1/46064…}] re-submitting command 6bd387b7c8370ea0 to Raft: reasonTicks
20:38:43.686813	1.000254	... [n1,s1,r22/1:/Table/51{-/1/46064…}] re-submitting command 6bd387b7c8370ea0 to Raft: reasonNewLeader
20:38:48.686696	4.999884	... [n1,s1,r22/1:/Table/51{-/1/46064…}] re-submitting command 6bd387b7c8370ea0 to Raft: reasonTicks
20:38:51.708141	3.021445	... [n1,s1,r22/1:/Table/51{-/1/46064…}] re-submitting command 6bd387b7c8370ea0 to Raft: reasonTicks
20:38:54.686814	2.978673	... [n1,s1,r22/1:/Table/51{-/1/46064…}] re-submitting command 6bd387b7c8370ea0 to Raft: reasonTicks
20:38:57.686569	2.999755	... [n1,s1,r22/1:/Table/51{-/1/46064…}] re-submitting command 6bd387b7c8370ea0 to Raft: reasonTicks
20:39:00.686685	3.000117	... [n1,s1,r22/1:/Table/51{-/1/46064…}] re-submitting command 6bd387b7c8370ea0 to Raft: reasonTicks
20:39:04.591983	3.905297	... [n1,s1,r22/1:/Table/51{-/1/46064…}] re-submitting command 6bd387b7c8370ea0 to Raft: reasonTicks
20:39:07.686560	3.094577	... [n1,s1,r22/1:/Table/51{-/1/46064…}] re-submitting command 6bd387b7c8370ea0 to Raft: reasonTicks
20:39:10.686591	3.000031	... [n1,s1,r22/1:/Table/51{-/1/46064…}] re-submitting command 6bd387b7c8370ea0 to Raft: reasonTicks
20:39:13.686486	2.999895	... [n1,s1,r22/1:/Table/51{-/1/46064…}] re-submitting command 6bd387b7c8370ea0 to Raft: reasonTicks
20:39:16.686506	3.000020	... [n1,s1,r22/1:/Table/51{-/1/46064…}] re-submitting command 6bd387b7c8370ea0 to Raft: reasonTicks
20:39:19.686564	3.000057	... [n1,s1,r22/1:/Table/51{-/1/46064…}] re-submitting command 6bd387b7c8370ea0 to Raft: reasonTicks
20:39:22.686767	3.000204	... [n1,s1,r22/1:/Table/51{-/1/46064…}] re-submitting command 6bd387b7c8370ea0 to Raft: reasonTicks
20:39:25.686768	3.000001	... [n1,s1,r22/1:/Table/51{-/1/46064…}] re-submitting command 6bd387b7c8370ea0 to Raft: reasonTicks
20:39:28.686542	2.999774	... [n1,s1,r22/1:/Table/51{-/1/46064…}] re-submitting command 6bd387b7c8370ea0 to Raft: reasonTicks
20:39:31.686622	3.000081	... [n1,s1,r22/1:/Table/51{-/1/46064…}] re-submitting command 6bd387b7c8370ea0 to Raft: reasonTicks
20:39:34.686803	3.000181	... [n1,s1,r22/1:/Table/51{-/1/46064…}] re-submitting command 6bd387b7c8370ea0 to Raft: reasonTicks
20:39:38.286771	3.599968	... [n1,s1,r22/1:/Table/51{-/1/46064…}] re-submitting command 6bd387b7c8370ea0 to Raft: reasonTicks
20:39:41.286600	2.999829	... [n1,s1,r22/1:/Table/51{-/1/46064…}] re-submitting command 6bd387b7c8370ea0 to Raft: reasonTicks
20:39:44.286666	3.000066	... [n1,s1,r22/1:/Table/51{-/1/46064…}] re-submitting command 6bd387b7c8370ea0 to Raft: reasonTicks
20:39:47.286766	3.000100	... [n1,s1,r22/1:/Table/51{-/1/46064…}] re-submitting command 6bd387b7c8370ea0 to Raft: reasonTicks
20:39:50.286792	3.000027	... [n1,s1,r22/1:/Table/51{-/1/46064…}] re-submitting command 6bd387b7c8370ea0 to Raft: reasonTicks
20:39:53.286625	2.999833	... [n1,s1,r22/1:/Table/51{-/1/46064…}] re-submitting command 6bd387b7c8370ea0 to Raft: reasonTicks
20:39:56.286569	2.999944	... [n1,s1,r22/1:/Table/51{-/1/46064…}] re-submitting command 6bd387b7c8370ea0 to Raft: reasonTicks
20:39:59.286654	3.000085	... [n1,s1,r22/1:/Table/51{-/1/46064…}] re-submitting command 6bd387b7c8370ea0 to Raft: reasonTicks
20:40:02.286580	2.999926	... [n1,s1,r22/1:/Table/51{-/1/46064…}] re-submitting command 6bd387b7c8370ea0 to Raft: reasonTicks
20:40:05.286747	3.000167	... [n1,s1,r22/1:/Table/51{-/1/46064…}] re-submitting command 6bd387b7c8370ea0 to Raft: reasonTicks
20:40:09.054490	3.767743	... [n1,s1,r22/1:/Table/51{-/1/46064…}] re-submitting command 6bd387b7c8370ea0 to Raft: reasonTicks
20:40:11.886739	2.832249	... [n1,s1,r22/1:/Table/51{-/1/46064…}] re-submitting command 6bd387b7c8370ea0 to Raft: reasonTicks
20:40:14.886869	3.000130	... [n1,s1,r22/1:/Table/51{-/1/46064…}] re-submitting command 6bd387b7c8370ea0 to Raft: reasonTicks
20:40:17.886763	2.999893	... [n1,s1,r22/1:/Table/51{-/1/46064…}] re-submitting command 6bd387b7c8370ea0 to Raft: reasonTicks
20:40:20.886843	3.000080	... [n1,s1,r22/1:/Table/51{-/1/46064…}] re-submitting command 6bd387b7c8370ea0 to Raft: reasonTicks
20:40:23.886791	2.999949	... [n1,s1,r22/1:/Table/51{-/1/46064…}] re-submitting command 6bd387b7c8370ea0 to Raft: reasonTicks
20:40:26.886790	2.999999	... [n1,s1,r22/1:/Table/51{-/1/46064…}] re-submitting command 6bd387b7c8370ea0 to Raft: reasonTicks
20:40:29.886545	2.999755	... [n1,s1,r22/1:/Table/51{-/1/46064…}] re-submitting command 6bd387b7c8370ea0 to Raft: reasonTicks
20:40:32.886697	3.000152	... [n1,s1,r22/1:/Table/51{-/1/46064…}] re-submitting command 6bd387b7c8370ea0 to Raft: reasonTicks
20:40:35.886799	3.000102	... [n1,s1,r22/1:/Table/51{-/1/46064…}] re-submitting command 6bd387b7c8370ea0 to Raft: reasonTicks
20:40:38.886703	2.999904	... [n1,s1,r22/1:/Table/51{-/1/46064…}] re-submitting command 6bd387b7c8370ea0 to Raft: reasonTicks
20:40:42.886595	3.999892	... [n1,s1,r22/1:/Table/51{-/1/46064…}] re-submitting command 6bd387b7c8370ea0 to Raft: reasonTicks
20:40:45.886765	3.000170	... [n1,s1,r22/1:/Table/51{-/1/46064…}] re-submitting command 6bd387b7c8370ea0 to Raft: reasonTicks
20:40:48.886894	3.000129	... [n1,s1,r22/1:/Table/51{-/1/46064…}] re-submitting command 6bd387b7c8370ea0 to Raft: reasonTicks
20:40:51.886490	2.999595	... [n1,s1,r22/1:/Table/51{-/1/46064…}] re-submitting command 6bd387b7c8370ea0 to Raft: reasonTicks
20:40:54.886752	3.000262	... [n1,s1,r22/1:/Table/51{-/1/46064…}] re-submitting command 6bd387b7c8370ea0 to Raft: reasonTicks
20:40:57.886699	2.999947	... [n1,s1,r22/1:/Table/51{-/1/46064…}] re-submitting command 6bd387b7c8370ea0 to Raft: reasonTicks
20:41:00.886786	3.000088	... [n1,s1,r22/1:/Table/51{-/1/46064…}] re-submitting command 6bd387b7c8370ea0 to Raft: reasonTicks
20:41:03.886584	2.999797	... [n1,s1,r22/1:/Table/51{-/1/46064…}] re-submitting command 6bd387b7c8370ea0 to Raft: reasonTicks
20:41:06.886760	3.000177	... [n1,s1,r22/1:/Table/51{-/1/46064…}] re-submitting command 6bd387b7c8370ea0 to Raft: reasonTicks
20:41:09.886651	2.999891	... [n1,s1,r22/1:/Table/51{-/1/46064…}] re-submitting command 6bd387b7c8370ea0 to Raft: reasonTicks
20:41:13.889940	4.003289	... [n1,s1,r22/1:/Table/51{-/1/46064…}] re-submitting command 6bd387b7c8370ea0 to Raft: reasonTicks
20:41:17.486541	3.596601	... [n1,s1,r22/1:/Table/51{-/1/46064…}] re-submitting command 6bd387b7c8370ea0 to Raft: reasonTicks
20:41:20.486646	3.000105	... [n1,s1,r22/1:/Table/51{-/1/46064…}] re-submitting command 6bd387b7c8370ea0 to Raft: reasonTicks
20:41:23.486523	2.999877	... [n1,s1,r22/1:/Table/51{-/1/46064…}] re-submitting command 6bd387b7c8370ea0 to Raft: reasonTicks
20:41:26.487009	3.000485	... [n1,s1,r22/1:/Table/51{-/1/46064…}] re-submitting command 6bd387b7c8370ea0 to Raft: reasonTicks
20:41:29.486691	2.999682	... [n1,s1,r22/1:/Table/51{-/1/46064…}] re-submitting command 6bd387b7c8370ea0 to Raft: reasonTicks
20:41:32.486715	3.000024	... [n1,s1,r22/1:/Table/51{-/1/46064…}] re-submitting command 6bd387b7c8370ea0 to Raft: reasonTicks
20:41:35.486677	2.999963	... [n1,s1,r22/1:/Table/51{-/1/46064…}] re-submitting command 6bd387b7c8370ea0 to Raft: reasonTicks
20:41:38.486789	3.000112	... [n1,s1,r22/1:/Table/51{-/1/46064…}] re-submitting command 6bd387b7c8370ea0 to Raft: reasonTicks
20:41:41.486596	2.999807	... [n1,s1,r22/1:/Table/51{-/1/46064…}] re-submitting command 6bd387b7c8370ea0 to Raft: reasonTicks
20:41:44.486912	3.000316	... [n1,s1,r22/1:/Table/51{-/1/46064…}] re-submitting command 6bd387b7c8370ea0 to Raft: reasonTicks
20:41:48.486873	3.999962	... [n1,s1,r22/1:/Table/51{-/1/46064…}] re-submitting command 6bd387b7c8370ea0 to Raft: reasonTicks
20:41:51.486843	2.999970	... [n1,s1,r22/1:/Table/51{-/1/46064…}] re-submitting command 6bd387b7c8370ea0 to Raft: reasonTicks
20:41:54.486738	2.999895	... [n1,s1,r22/1:/Table/51{-/1/46064…}] re-submitting command 6bd387b7c8370ea0 to Raft: reasonTicks
```

So range 22 isn't making any progress.

**Comment by petermattis** [comment]
The `/_status/raft` page shows that range 22 is in a bit of a confused state. Three nodes think that range is composed of replicas on nodes 8, 9 and 10. But one node thinks it has replicas on nodes 1, 8, and 9. Note that the above `/debug/requests` dump was from node 1. So some commands on node 1 are stuck forever trying to refresh proposals on range 22. But it seems like range 22 no longer has a replica on node 1 (in the range descriptor).

**Comment by a-robinson** [comment]
I'm completely unfamiliar with how this part of the system works, so I have some dumb questions if you don't mind sparing a minute: Isn't `reasonTicks` a really weird error to be getting from a replica that's unable to communicate with the raft group that it thinks it's a part of? Why isn't the replica finding out that it's no longer a valid member of the raft group and should be GC'ed? How is that supposed to work?

**Comment by petermattis** [comment]
`reasonTicks` is what happens when a command is proposed but never executes. The usual reason for this reason is that the replica doesn't know the leader for the range. I don't know why this replica isn't being GC'd. Usually this happens by it sending a message to another member of the group and getting told it isn't a member. There is something quite basic that is broken here.

**Comment by bdarnell** [comment]
@tamird I don't think the deadlock tag would help with this particular deadlock, unless it's smarter than I'm imagining. The ordering violation is between a regular mutex and a `chan struct{}` being used as a semaphore. I don't think it would be able to understand that pattern (but maybe there are other more conventional deadlocks that it could pick up on)

@a-robinson 
> Isn't reasonTicks a really weird error to be getting from a replica that's unable to communicate with the raft group that it thinks it's a part of?

No, that's exactly what you'd expect of a node that can't communicate with its raft group. `reasonTicks` is the retry process for raft proposals that get eaten by the network. The question is why the nodes it's talking to aren't sending back errors telling it to clean up its garbage replica. I think the most likely answer is that a deadlock on raftMu is still present, so this is a symptom of a healthy node trying to talk to a node whose raftMu is deadlocked.

**Comment by mberhault** [comment]
wiped and restarted cyan with fc005be. No issues after starting all block_writer tasks at once (shortly after startup).
However, there's massive unbalance in lease holders, with the 6 nodes having 32, 3, 2, 1, 0, 0 leases.

**Comment by petermattis** [comment]
Lease rebalancing is still disabled by default, so that behavior is
expected. I explicitly enabled it on blue.

**Comment by petermattis** [comment]
I'm going to try bisecting to figure out when the problem on delta started occurring.

**Comment by petermattis** [comment]
`beta-20161103` is good, `beta-20161110` is bad. Only 71 commits [6 steps] to go.

**Comment by petermattis** [comment]
And the finger points to 7250330136f2cb025d40f9d9b7754260854f046f:

```
commit 7250330136f2cb025d40f9d9b7754260854f046f
Author: Ben Darnell <ben@cockroachlabs.com>
Date:   Sat Nov 5 11:20:11 2016 +0800

    storage: Don't add NoopRequests to the command queue

    Doing so was creating unnecessary contention in the command queue for
    the empty key.
```

I didn't see that coming.

**Comment by petermattis** [comment]
That change looks innocuous. Could something be relying on the contention on the empty key? Paging @bdarnell.

**Comment by petermattis** [comment]
Uh-oh, delta wedged again. This time it happened after running fine for 15 min.  Perhaps 7250330 only uncovered another lurking problem. That would make more sense as I can't see how skipping NoopRequests for the command queue could possibly cause a problem.

**Comment by petermattis** [comment]
Seems like there is some problem related to splits. In the graph below the goroutine spike corresponds to a series of splits being performed. And the earlier delta wedgings were all occurring during the initial split of the `datablocks.blocks` table.

<img width="932" alt="screen shot 2016-11-19 at 3 58 12 pm" src="https://cloud.githubusercontent.com/assets/6201335/20458353/1b399f18-ae71-11e6-86f3-a25323c866b7.png">

**Comment by petermattis** [comment]
I manually applied the change to `replica.go` from 7250330 on top of `beta-20161103` and it seems to be running fine. This supports the theory that 7250330 is not to blame but merely uncovered a preexisting condition.

**Comment by tamird** [comment]
Yikes, so now to bisect again with 7250330 applied to each commit?

**Comment by petermattis** [comment]
Something like that.

**Comment by petermattis** [comment]
Actually, it turns out that `beta-20161103` was problematic too. The issue on delta only shows up (quickly) with `block_writer -batch 10`. Bisect back the finger is now pointing to 58c601243091caa8e5624da72deb6766ec155d88. Applying the change to `replica.go` from 7250330 on top of that commit and `delta` wedges up within a few minutes. The previous commit (4f9f35549eb239a6f9989cdad3a55b17d9b9d0a1) looks fine.

58c601243091caa8e5624da72deb6766ec155d88 split the command queue into local and global keys. Seems like that fouled something up for splits, but only when there are multi-range transactions active.

**Comment by bdarnell** [comment]
It makes much more sense that 58c60124 would be the culprit, instead of 7250330. The NoopRequest contention must have been serializing things enough that we never hit whatever interaction is causing problems. 

Do you have a local repro for this or are you bisecting by deploying to delta? 

The batch_writer transactions alone should not have been affected by 58c60124 - they only touch global keys. Maybe it's an interaction between the batch_writer transactions and split/rebalance transactions, which mix global and local keys. The global keys touched by the two types of transactions should also be disjoint, but maybe there are issues with the "covering optimization", which turns multiple point operations into a span operation. 

Two simple experiments to try would be turning off the covering optimization for the global queue ([here](https://github.com/cockroachdb/cockroach/commit/58c601243091caa8e5624da72deb6766ec155d88#diff-0f4e61b240de001f1b97a87fc31af043R592)), or turning it *on* for the local queue (the covering optimization was always enabled prior to this commit, so maybe the non-optimized version has a bug).

**Comment by bdarnell** [comment]
@petermattis Try the change in #10854 and see if that fixes it.

**Comment by petermattis** [comment]
I only ever reproduced this problem on delta and thus bisected there. 

#10854 did not fix the problem (see delta graphs between 10:38 and 10:42).

I suspect that disabling the covering optimization has a bug, so that's what I'm trying next (enabling it for the local command queue).

**Comment by petermattis** [comment]
Enabling the covering optimization for the local command queue didn't help either. See delta between 10:45 and 10:47. I also tried disabling usage of the local command queue completely (by always adding to `spansGlobal`), but that didn't work either. Lastly, I tried disabling the covering optimization for both the local and global command queues. Finally, we have a winner! I've looked at the code and nothing obvious jumps out.

This is the diff on top of #10854:

```diff
diff --git a/pkg/storage/replica.go b/pkg/storage/replica.go
index 858615f..120ff6b 100644
--- a/pkg/storage/replica.go
+++ b/pkg/storage/replica.go
@@ -624,7 +624,7 @@ func (r *Replica) initLocked(
        }

        r.cmdQMu.Lock()
-       r.cmdQMu.global = NewCommandQueue(true /* optimizeOverlap */)
+       r.cmdQMu.global = NewCommandQueue(false /* optimizeOverlap */)
        r.cmdQMu.local = NewCommandQueue(false /* !optimizeOverlap */)
        r.cmdQMu.Unlock()
```

**Comment by petermattis** [comment]
I might have spoken too soon (again). Disabling the covering optimization doesn't appear to completely fix the problem. Sometimes a fresh cluster works fine and other times it wedges up exactly as before.

**Comment by petermattis** [comment]
I finally seem to have reliably reproduced this badness locally.

```
~ killall -9 cockroach; rm -fr cockroach-data/ && roachdemo -n 3 -- --logtostderr
...
~ block_writer -concurrency 64 -tolerate-errors -batch 10 'postgresql://root@localhost:26257/photos?sslmode=disable'
...
~ block_writer -concurrency 64 -tolerate-errors -batch 10 'postgresql://root@localhost:26259/photos?sslmode=disable'
...
~ block_writer -concurrency 64 -tolerate-errors -batch 10 'postgresql://root@localhost:26261/photos?sslmode=disable'
```

So 3 different block_writers hitting each of the 3 nodes in the local cluster. Both the `-concurrency` and `-batch` flags are needed to reproduce, though I'm not sure why yet.

**Comment by petermattis** [comment]
There seems to be some deadlock situation with intent resolution. I noticed a lot of `failed to resolve intents` messages in the logs when the error occurred locally. If I bump `intentResolverTaskLimit` from 100 to 1000, the problem goes away. This would also help explain why the `-batch 10` setting for block_writer is needed. With the default `-batch 1`, block_writer doesn't creating any intents.

**Comment by petermattis** [comment]
Consider the following call stack:

```
    stop      stopper.go:296          (*Stopper).RunLimitedAsyncTask(#237, #108, #201, #593, #200, *, 0, 0)
    storage   intent_resolver.go:385  (*intentResolver).processIntentsAsync(#455, *, *, 0x1, 0x1)
    storage   replica_proposal.go:621 (*Replica).handleLocalProposalData(*, #110, *, #166, 0x1, 0, 0, 0, 0, 0, ...)
    storage   replica_proposal.go:697 (*Replica).handleProposalData(*, #110, *, #166, 0x1, *, 0x8, 0, 0, 0, ...)
    storage   replica.go:3215         (*Replica).processRaftCommand(*, #110, *, *, 0x8, *, #166, 0x1, *, *, ...)
    storage   replica.go:2373         (*Replica).handleRaftReadyRaftMuLocked(*, 0, 0, 0, 0, 0, 0, 0, 0, 0, ...)
    storage   replica.go:2201         (*Replica).handleRaftReady(*, 0, 0, 0, 0, 0, 0, 0, 0, 0, ...)
    storage   store.go:3201           (*Store).processReady(#544, *)
    storage   scheduler.go:230        (*raftScheduler).worker(#529, #237)
    storage   scheduler.go:181        (*raftScheduler).Start.func2()
    stop      stopper.go:196          (*Stopper).RunWorker.func1(#237, *)
```

Intent resolution is run while holding `Replica.raftMu`. I haven't completely identified badness here, but that doesn't seem good.

**Comment by petermattis** [comment]
Btw, the current setting of `intentResolverTaskLimit` (100) also explains why I wasn't reproducing the problem earlier. In my previous attempts at reproduction I was using the default `block_writer` concurrency (`2 * num-cpu == 8` on my laptop). We should be wary of any use of `RunLimitedAsyncTask`.
